design/deploy-steps.md

```diff
@@ -0,0 +1,149 @@
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer ([coreos-installer][coreos-installer]).
+
+[coreos-installer]: https://github.com/coreos/coreos-installer
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- A way to provide a link to the Ironic agent image per host.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: CustomDeploy
+
+When `CustomDeploy` is set, BMO will set the new [custom-agent deploy
+interface][custom-agent] on the node so that the default deploy steps do not
+apply. Then we request a deploy step with the name provided in the
+`CustomDeploy.Method` field. This step has to be provided in the agent image
+and is responsible for the whole deployment process.
+
+The `Image.URL` will no longer be required since some images (such as CoreOS)
+already include the root filesystem.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
+in the future. The `UserData` and `NetworkData` will be available to a deploy
+step implementation as part of the Node's `instance_info["configdrive"]`.
+Any required customization is expected to be provided via `UserData` and
+`NetworkData` or be built into the image itself.
+
+The expected workflow is as follows:
+
+1. A user sets `CustomDeploy.Method = "step_name"` and optionally `Image.URL`,
+   `UserData` and `NetworkData`.
+2. BMO sets `Node.DeployInterface = "custom-deploy"` and requests deployment
+   with a custom deploy step `step_name`.
+3. Ironic boots the agent (if not booted for inspection/cleaning already).
+4. Ironic requests the agent to execute the deploy step `step_name` (it is
+   an error to provide a non-existing step), providing it the complete Node
+   object, including `configdrive`.
+5. Ironic waits for the step to finish.
+6. Ironic configures the Node's boot device and clean ups the deployment
+   environment.
+7. Ironic reboots the node into the instance.
+
+On failure (including unsupported deploy step) the node will be moved to the
+`deploy failed` state, and the `LastError` field will be populated.
```

Comment:
That's the Ironic behavior. From a BMO point of view, the BareMetalHost will remain in the `Provisioning` state, with the error fields set

Reply:
Right, I think it's useful in general in such kind of design documents to hilight  the expected BMO/Ironic behavior, considering that they have two different state machines.